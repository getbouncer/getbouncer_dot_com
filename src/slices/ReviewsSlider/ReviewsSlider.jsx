import React, { useCallback, useEffect, useState, cloneElement } from 'react'
import PropTypes from 'prop-types'
import { useEmblaCarousel } from 'embla-carousel/react'
import styled from '@emotion/styled'
import useInterval from '~hooks/useInterval'
import Container from '~globals/Container'
import arrowSlider from '~images/arrow-slider.svg'
import {
  InnerContainer,
  SliderContainer,
  Slide,
  Text,
  Author,
  DotPrev,
  DotNext,
  LogosContainer,
  Logo,
  Wrapper,
} from './styles'

const Viewport = styled(({ component, ...props }) => cloneElement(component, props))`
  overflow: hidden;

  &.is-draggable {
    cursor: move;
    cursor: grab;
  }

  &.is-dragging {
    cursor: grabbing;
  }
`

const ReviewsSlider = ({ input }) => {
  const { items } = input
  const [EmblaCarouselReact, embla] = useEmblaCarousel({ loop: true })
  const scrollTo = useCallback(index => embla.scrollTo(index), [embla])
  const scrollPrev = useCallback(() => embla.scrollPrev(), [embla])
  const scrollNext = useCallback(() => embla.scrollNext(), [embla])
  const [isRunning, setIsRunning] = useState(false)

  useInterval(
    () => {
      scrollNext()
    },
    isRunning ? null : 6000,
  )

  useEffect(() => {
    if (embla) {
      embla.on('select', () => {
        setIsRunning(true)
      })

      embla.on('settle', () => {
        setIsRunning(false)
      })
    }
  }, [embla])

  return (
    <Wrapper>
      <Container isLarge>
        <InnerContainer>
          <Viewport component={<EmblaCarouselReact autoplay delayLength={4000} />}>
            <SliderContainer>
              {items.map((item, index) => (
                // eslint-disable-next-line react/no-array-index-key
                <Slide key={index}>
                  <img src={item.company_logo.url} alt={item.company_logo.alt} />
                  <Text>{item.review_text.text}</Text>
                  <Author>{item.review_author}</Author>
                </Slide>
              ))}
            </SliderContainer>
          </Viewport>
          <DotPrev onClick={scrollPrev}>
            <img src={arrowSlider} alt="Previous arrow" />
          </DotPrev>
          <DotNext onClick={scrollNext}>
            <img src={arrowSlider} alt="Next arrow" />
          </DotNext>
        </InnerContainer>
        <LogosContainer>
          {items.map((item, index) => (
            // eslint-disable-next-line react/no-array-index-key
            <Logo key={index} onClick={() => scrollTo(index)}>
              <img
                src={item.company_logo.url}
                alt={item.company_logo.alt}
                className={embla && embla.selectedScrollSnap() === index ? 'active' : undefined}
              />
            </Logo>
          ))}
        </LogosContainer>
      </Container>
    </Wrapper>
  )
}

ReviewsSlider.propTypes = {
  input: PropTypes.shape().isRequired,
}

export default ReviewsSlider
