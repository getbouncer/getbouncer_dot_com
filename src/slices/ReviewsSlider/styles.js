import styled from '@emotion/styled'
import { breakpoints, theme } from '~styles'

export const Wrapper = styled.section`
  & > div {
    @media ${breakpoints.smartphone} {
      padding: 0;
    }
  }
`

export const InnerContainer = styled.div`
  background-color: ${theme.colors.blue};
  padding: 85px 0 120px;
  position: relative;

  @media ${breakpoints.smartphone} {
    padding: 55px 32px;
  }
`

export const SliderContainer = styled.ul`
  display: flex;
  will-change: transform;
`

export const Slide = styled.li`
  color: white;
  flex: 0 0 auto;
  position: relative;
  text-align: center;
  width: 100%;

  img {
    height: 30px;
    filter: brightness(0) invert(1);
  }
`

export const Icon = styled.li`
  cursor: pointer;

  img {
    height: 45px;
  }

`

export const Text = styled.p`
  font-size: 1.75rem;
  line-height: 2.6rem;
  margin: 66px auto 36px;
  max-width: 700px;

  @media ${breakpoints.smartphone} {
    font-size: 1.15rem;
    line-height: 1.8rem;
  }
`

export const Author = styled.span`
  color: #e9ecfe;
`

const Dot = styled.div`
  align-items: center;
  background-color: white;
  border-radius: 100%;
  cursor: pointer;
  display: flex;
  height: 48px;
  justify-content: center;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  width: 48px;

  @media ${breakpoints.smartphone} {
    display: none;
  }
`

export const DotPrev = styled(Dot)`
  left: 50px;
`

export const DotNext = styled(Dot)`
  right: 50px;

  img {
    transform: rotate(180deg);
  }
`

export const LogosContainer = styled.ul`
  align-items: center;
  display: flex;
  justify-content: center;
  margin: 35px 0 150px;
`

export const Logo = styled.li`
  cursor: pointer;

  img {
    height: 30px;

    &:not(.active) {
      filter: brightness(0);
    }
  }

  &:not(:last-child) {
    margin-right: 100px;
  }
`
