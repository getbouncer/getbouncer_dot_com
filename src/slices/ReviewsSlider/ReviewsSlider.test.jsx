import React from 'react'
import renderer from 'react-test-renderer'
import ReviewsSlider from './ReviewsSlider'

/* eslint-disable no-undef */

describe('ReviewsSlider', () => {
  it('renders correctly', () => {
    const data = {
      items: [
        {
          company_logo: {
            alt: 'company 1 logo',
            url: 'https://www.example.com/company_1_logo',
          },
          review_text: {
            text: 'review 1',
          },
          review_author: 'review 1 author',
        },
        {
          company_logo: {
            alt: 'company 2 logo',
            url: 'https://www.example.com/company_2_logo',
          },
          review_text: {
            text: 'review 2',
          },
          review_author: 'review 2 author',
        },
      ],
    }
    const tree = renderer
      .create(<ReviewsSlider input={data} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
