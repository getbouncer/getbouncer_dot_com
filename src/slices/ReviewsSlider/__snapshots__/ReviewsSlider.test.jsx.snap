// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewsSlider renders correctly 1`] = `
<section
  className="css-fzaw5e"
>
  <div
    className="css-sdcyqc"
  >
    <div
      className="css-faouk5"
    >
      <div
        className="css-1e2ahhp"
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <ul
          className="css-erihxk"
        >
          <li
            className="css-5izmqs"
          >
            <li
              className="css-nrckiy"
            >
              <img
                alt="company 1 logo"
                src="https://www.example.com/company_1_logo"
              />
            </li>
            <p
              className="css-11njosq"
            >
              review 1
            </p>
            <span
              className="css-17odb7k"
            >
              review 1 author
            </span>
          </li>
          <li
            className="css-5izmqs"
          >
            <li
              className="css-nrckiy"
            >
              <img
                alt="company 2 logo"
                src="https://www.example.com/company_2_logo"
              />
            </li>
            <p
              className="css-11njosq"
            >
              review 2
            </p>
            <span
              className="css-17odb7k"
            >
              review 2 author
            </span>
          </li>
        </ul>
      </div>
      <div
        className="css-f2awql"
        onClick={[Function]}
      >
        <img
          alt="Previous arrow"
          src="test-file-stub"
        />
      </div>
      <div
        className="css-1m70fi3"
        onClick={[Function]}
      >
        <img
          alt="Next arrow"
          src="test-file-stub"
        />
      </div>
    </div>
    <ul
      className="css-znwfed"
    >
      <li
        className="css-qwhy2v"
        onClick={[Function]}
      >
        <img
          alt="company 1 logo"
          src="https://www.example.com/company_1_logo"
        />
      </li>
      <li
        className="css-qwhy2v"
        onClick={[Function]}
      >
        <img
          alt="company 2 logo"
          src="https://www.example.com/company_2_logo"
        />
      </li>
    </ul>
  </div>
</section>
`;
